File: testTask.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"strconv"
)

type Product struct {
	DisplayedName struct {
		DisplayedName struct {
			Value       []string `json: "value"`
			Description string `json: "description"`
		} `json: "displayedName"`
	} `json: "displayedName"`
	Stock struct {
		Stocks map[string]map[string]string `json:"stocks"`
	} `json: "stock"`
}

func (p Product) getValueOfDisplayedName() string {
	return p.DisplayedName.DisplayedName.Value[0]
}

func (p Product) getArrayOfShopNumbers(region string) []string {
	result := make([]string, 0)
	for key, val := range p.Stock.Stocks[region] {
		cmpVal, _ := strconv.Atoi(val)
		if cmpVal > 0 {
			result = append(result, key)
		}
	}
	return result
}

func (p Product) getMaxOfStocksAt(region string) (maxVal int, resultKey string) {
	for key, val := range p.Stock.Stocks[region] {
		cmpVal, _ := strconv.Atoi(val)
		if cmpVal > maxVal {
			maxVal = cmpVal
			resultKey = key
		}
	}
	return
}

func main() {
	data, err := ioutil.ReadFile("file.json")
	if err != nil {
		fmt.Print(err)
	}

	var prod Product//map[string]interface{}//Product

	_ = json.Unmarshal(data, &prod)

	fmt.Println(prod.getValueOfDisplayedName())
	fmt.Println(prod.getArrayOfShopNumbers("34"))
	fmt.Println(prod.getMaxOfStocksAt("34"))

}

File: testTask.js
var obj = JSON.parse(`{
  "displayedName": {
      "displayedName": {
          "value": [
              "Профиль маячковый ПВХ 10 мм L3м"
          ],
          "description": "Полное наименование товара для клиента"
          }
      },
  "stock": {
      "stocks": {
              "34": {
              "2": "35",
              "3": "42",
              "4": "58",
              "5": "57",
              "6": "112",
              "20": "51",
              "22": "78",
              "26": "34",
              "32": "22",
              "35": "358",
              "40": "28",
              "43": "68",
              "45": "58",
              "49": "31",
              "51": "29",
              "56": "42",
              "62": "26",
              "64": "0",
              "65": "57",
              "86": "15",
              "114": "41",
              "117": "46",
              "143": "46",
              "162": "4",
              "171": "0",
              "176": "12"
          }
      }
  }
}`);

function getValueOfDisplayedName(SomeObj) {
  return SomeObj.displayedName.displayedName.value[0]
}

function getArrayOfShopNumbers(SomeObj, region) {
  let result = []
  const objOfShops = SomeObj.stock.stocks[region]
  for (key in objOfShops) {
    if (objOfShops[key] > 0) {
      result.push(key)
    }
  }
  return result
}


function getMaxOfStocksAt(SomeObj, region) {
  const objOfShops = SomeObj.stock.stocks[region]
  let maxVal = -1
  let resultKey = -1
  for(key in objOfShops) {
    maxVal = Math.max(objOfShops[key], maxVal)
    if( maxVal != objOfShops[resultKey]) {
      resultKey = key
    }
  }
  return {maxVal, resultKey}
}
